pyroomacoustics/libroom_src/utility.cpp

```diff
@@ -0,0 +1,390 @@
+			 float m = eq[0];
+			 float c = eq[1];
+			 
+			 float A = m*m +1;
+			 float B = 2 * (m*c - m*q - p);
+			 float C = q*q - radius*radius + p*p - 2*c*q + c*c;
+			 
+			 Vector2f xs = solve_quad(A,B,C);
+			 Vector2f ys = Vector2f(m*xs[0] + c, m*xs[1] + c);
+			 
+			 result1 = Vector2f(xs[0], ys[0]);		 
+			 result2 = Vector2f(xs[1], ys[1]);		 
+		 }
+		 
+		 // Now we must return only the closest intersection point 
+		 // to the start of the segment
+
+		 if ( (start-result2).norm() < (start-result1).norm() ){
+			 return result2;
+		 }
+		 return result1;		 
+	 }
+	 
+	 
+	 if (start.size() == 3){
+		 
+		 Vector3f result1;
+		 Vector3f result2;
+		 
+		 // Here we are going to follow this formula :
+		 // https://en.wikipedia.org/wiki/Line%E2%80%93sphere_intersection
+		 // With respect to the formula given on this page :
+		 // o = start
+		 // c = center
+		 // r = radius
+		 
+		 // We can check the results with this app :
+		 // http://www.ambrsoft.com/TrigoCalc/Sphere/SpherLineIntersection_.htm
+
+		Vector3f l = (end-start).normalized();
+		
+        // We are searching d such that a point x on the sphere is also on the line
+        // ie : x = o + k*l  (recall l has unit norm)		
+		
+		 Vector3f o_c = start-center;
+		 float ocl_dot = o_c.dot(l);
+		 float delta = ocl_dot*ocl_dot - o_c.norm()*o_c.norm() + radius*radius;
+		 
+		 if (delta > 0){
+			 
+			 result1 = start - l*(ocl_dot + sqrt(delta));
+			 result2 = start - l*(ocl_dot - sqrt(delta));
+			 
+			 if ( (start-result2).norm() < (start-result1).norm() ){
+				 return result2;
+			 }
+			 return result1;
+		 }
+		 
+		 // As this function is only called when there is an intersection,
+		 // delta should not be negative, unless because of rounding errors.
+		 // If this happens, we consider the delta to be 0
+		 
+		 return start - l*ocl_dot;
+		 
+		 
+	 }
+	 
+	std::cerr << "Error : the vectors should be 2D or 3D" << s_center <<std::endl;
+	throw std::exception();
+}
+
+
+void update_travel_time(float &travel_time,
+						float hop_length,
+						float sound_speed){
+							
+	/* This function updates the travel time according to the newly 
+	 * travelled distance*/
+	 
+	travel_time = travel_time + hop_length/sound_speed;	
+}
+
+
+void update_energy_wall(float &energy,
+						const Wall &wall){
+							
+	/* This function updates the travel time according to the newly 
+	 * travelled distance*/
+	 
+	energy = energy * sqrt(1 - wall.absorption);
+}
+
+
+float compute_scat_energy(float energy,
+						  float scat_coef,
+						  const Wall &wall, 
+						  const VectorXf &start,
+						  const VectorXf &hit_point,
+						  const VectorXf &mic_pos){
+	
+	VectorXf n = wall.normal;
+	
+	// Make sure the normal points inside the room
+	
+	if (angle_between(start-hit_point, n) > M_PI_2){
+		n = (-1)*n;
+	}
+	
+	return energy * scat_coef * cos(angle_between(n, mic_pos-hit_point));
+}
+
+
+void append(float travel_time, float energy, std::vector<entry> &output){
```

Comment:
This function is not needed, the STL vector class already takes care of resizing the allocated memory when necessary.

Reply:
As we have a lot of entries, I wrote this function to avoid several useless resizing (ie it is non-sense to have a vector with an initial capacity of 2 when we can directly set it higher right ?)

So if I remove this function, there will be more automatic calls to the resize function and thus more copies of the data, is that really something we want ?

Reply:
The `vector` class has already constant amortized cost for the `push_back` function. Internally it does exactly what you are doing. When the vector is too small, it will double the size of the allocated storage (there is typically more storage allocated than the current size of the vector).

From [this reference](http://www.cplusplus.com/reference/vector/vector/):
```
Internally, vectors use a dynamically allocated array to store their elements. This array may need to be reallocated in order to grow in size when new elements are inserted, which implies allocating a new array and moving all elements to it. This is a relatively expensive task in terms of processing time, and thus, vectors do not reallocate each time an element is added to the container.
```

It is indeed not a bad idea to pre-allocate some large chunk of memory, but the way you are currently doing it is also update the `size` and filling with zeros. Instead you should use the `reserve` function to only change the size of the allocated memory (not of the vector).

In fact vectors use contiguous storage, which might not be needed in our case. It might be more efficient to use a [`forward_list`](http://www.cplusplus.com/reference/forward_list/forward_list/).

Reply:
Looking at [this page](https://github.com/pybind/pybind11/blob/master/include/pybind11/stl.h), it seems that forward lists are not supported by pybind11

Do you think it would be worth to try with basic (double linked) lists ?

Reply:
Yes, doubly-linked list consume a little bit more memory (one extra pointer), but I don't think this should be a problem. Let's try with this.